core/src/main/kotlin/io/customer/android/sdk/CustomerIO.kt

```diff
@@ -0,0 +1,79 @@
+package io.customer.android.sdk
+
+import androidx.annotation.VisibleForTesting
+import io.customer.android.core.di.SDKComponent
+import io.customer.base.internal.InternalCustomerIOApi
+
+// TODO: Move this class and its dependencies (CustomerIOInstance) to the correct package.
+// We need to move this class to the right package to avoid breaking imports for the users of the SDK.
+// We have placed the class in the wrong package for now to avoid breaking the build.
+// Once old implementations are removed, we can move the class to the correct package.
+
+/**
+ * Welcome to the Customer.io Android SDK!
+ * This class is where you begin to use the SDK.
+ * You must have an instance of `CustomerIO` to use the features of the SDK.
+ * Create your own instance using
+ * ```
+ * with(CustomerIO.Builder(appContext: Application context, cdpApiKey = "XXX")) {
+ *   setLogLevel(...)
+ *   addCustomerIOModule(...)
+ *   build()
+ * }
+ * ```
+ * It is recommended to initialize the client in the `Application::onCreate()` method.
+ * After the instance is created you can access it via singleton instance: `CustomerIO.instance()` anywhere,
+ */
+class CustomerIO private constructor(
+    implementation: CustomerIOInstance
+) : CustomerIOInstance by implementation {
+    companion object {
+        /**
+         * Singleton instance of CustomerIO SDK that is created and set using the provided implementation.
+         */
+        @Volatile
+        private var instance: CustomerIO? = null
+
+        /**
+         * Returns the instance of CustomerIO SDK.
+         * If the instance is not initialized, it will throw an exception.
+         * Please ensure that the SDK is initialized before calling this method.
+         */
+        @JvmStatic
+        fun instance(): CustomerIO {
+            return instance ?: throw IllegalStateException(
+                "CustomerIO is not initialized. CustomerIO.Builder::build() must be called before obtaining SDK instance."
+            )
+        }
+
+        /**
+         * Creates and sets new instance of CustomerIO SDK using the provided implementation.
+         * If the instance is already initialized, it will log an error and skip the initialization.
+         * This method should be called only once during the application lifecycle using the provided builder.
+         */
+        @Synchronized
+        @InternalCustomerIOApi
+        fun createInstance(
+            implementation: CustomerIOInstance
+        ): CustomerIO {
+            val existingInstance = instance
+            if (existingInstance != null) {
+                SDKComponent.logger.error("CustomerIO instance is already initialized, skipping the initialization.")
+                return existingInstance
+            }
+            return CustomerIO(implementation = implementation).apply {
+                instance = this
+            }
+        }
```

Comment:
small performance improvement we could do is instead of marking the whole method with `@Synchronized`, we could just skip the null check as it doesn't require synchronization.  Synchronization is only necessary when the instance is being created

```
fun createInstance(implementation: CustomerIOInstance): CustomerIO {
            return instance ?: synchronized(this) {
                instance ?: CustomerIO(implementation).also {
                    instance = it
                }
            }
        }
 ```
 
 this is without log, but you can make a version with logger too.

Reply:
Good call, I didn't realize that will acquire lock even if the instance is already created, updating.

Reply:
I realized that's incorrect. We need to make sure we return instance from lock too. Example use case

- Two threads try to access instance together
- Both of them gets `null`
- One of them enters the lock to create and set new instance
- The other keeps waiting, but has already passed `null` check
- The first one leaves and second one enters the lock to create and set new instance resulting in two separate instances

So `@Synchronized` and `synchronized` block does the same thing here. I'll stick with `@Synchronized` as it is easier to read.

Reply:
I am okay with `@Synchronized`